src/main/cpp/native_glue.cc

```diff
@@ -19,140 +19,89 @@
-    for(size_t i = 0; i < len; i++) {
-        if(unsignedInput[i] >= 0xF0) {
-            // 0xF0 and above marks 4-byte sequences
-            extraBytes += 2;
-
-            // 3 + 1 from next loop iteration = 4 bytes advance
-            i += 3;
-        }
-    }
-
-    if(extraBytes == 0) {
-        // No modifications necessary
-        *output = nullptr;
-        return len;
+std::string litecore::jni::JstringToUTF8(JNIEnv *env, jstring jstr) {
+    jsize len = env->GetStringLength(jstr);
+    if (len < 0) {
+        C4Error error = {LiteCoreDomain, kC4ErrorInvalidParameter, 0};
+        throwError(env, error);
+        return std::string();
     }
 
-    size_t newStrLen = len + extraBytes;
-    char* newBytes = (char *)malloc(newStrLen);
-    if(newBytes == nullptr) {
-        *output = nullptr;
-        return -1;
+    const jchar *chars = env->GetStringChars(jstr, nullptr);
+    if (chars == nullptr) {
+        C4Error error = {LiteCoreDomain, kC4ErrorMemoryError, 0};
+        throwError(env, error);
+        return std::string();
     }
 
-    int offset = 0;
-    for(size_t i = 0; i < len; i++) {
-        if(unsignedInput[i] >= 0xF0) {
-            UTF8CharToModifiedUTF8(input + i, newBytes + i + offset);
-            i += 3;
-            offset += 2;
-        } else {
-            newBytes[i+offset] = unsignedInput[i];
-        }
-    }
+    try {
+        auto str = std::wstring_convert<std::codecvt_utf8_utf16<char16_t>, char16_t>()
+                .to_bytes(reinterpret_cast<const char16_t *>(chars), reinterpret_cast<const char16_t *>(chars + len));
 
-    *output = newBytes;
-    return newStrLen;
-}
+        env->ReleaseStringChars(jstr, chars);
 
-void litecore::jni::ModifiedUTF8CharToUTF8(char *input) {
-    char c = input[0];
-    char c1 = input[1] & 0x3F;
-    char c2 = input[2] & 0x3F;
-    char d = input[3];
-    char d1 = input[4] & 0x3F;
-    char d2 = input[5] & 0x3F;
-
-    int surrogate[2] = { ((c & 0x0F) << 12) | (c1 << 6) | c2, ((d & 0x0F) << 12) | (d1 << 6) | d2 };
-    int codePoint = ((surrogate[0] - 0xD800) << 10) + (surrogate[1] - 0xDC00) + 0x10000;
-
-    input[0] = 0xF0 | (codePoint  >> 18);
-    input[1] = 0x80 | ((codePoint >> 12) & 0x3F);
-    input[2] = 0x80 | ((codePoint >> 6) & 0x3F);
-    input[3] = 0x80 | ((codePoint & 0x3F));
-}
-
-void litecore::jni::ModifiedUTF8ToUTF8(char *input) {
-    size_t len = 0;
-    size_t i = 0;
-    bool needsModify = false;
-    for(i = 0; input[i] != 0; i++) {
-        if(input[i] == '\xed') {
-            // According to https://docs.oracle.com/javase/1.5.0/docs/guide/jni/spec/types.html#wp16542
-            // modified UTF-8 for codepoints > FFFF will always start with 0xED
-            needsModify = true;
-            ModifiedUTF8CharToUTF8(&input[i]);
-            i += 5;
-        }
+        return str;
     }
+    catch (const std::exception &x) { }
 
-    if(!needsModify) {
-        return;
-    }
+    env->ReleaseStringChars(jstr, chars);
```

Comment:
This cleanup code should go inside the `catch` block to make it clearer that it's part of exception handling ... I was confused by it for a minute.

Reply:
fixed